feat: Introducing deployment plans (#333)
hirosystems/clarinet: README.md

Make sure that you have a working installation of Docker running locally.

### Interacting with contracts deployed on Mainnet

Composition and interactions between protocols and contracts are one of the key innovations in blockchains. Clarinet was designed to handle this sort of interactions.

> Before refering to contracts deployed on Mainnet, they should be explicitily be listed as a `requirement` in the manifest `Clarinet.toml`:

Reviewer: ```suggestion
Before referring to contracts deployed on Mainnet, they should be explicitily be listed as a `requirement` in the manifest `Clarinet.toml`:
```

> Before refering to contracts deployed on Mainnet, they should be explicitily be listed as a `requirement` in the manifest `Clarinet.toml`:

Reviewer: Maybe also mention that `clarinet contracts requirement <contract-id>` can be used to add this.

Author: Indeed, I forgot that we had this command. I revisited it the approach, and went with:

```bash
clarinet requirements add <contract-id>
```

so we have more room for other eventual future requirements related subcommands (remove, remap, ?).

```toml
[project]
name = "my-project"
requirements = [

>   "SP2KAF9RF86PVX3NEE27DFV1CQX0T4WGR41X3S45C.bitcoin-whales.payout"

Reviewer: ```suggestion
  "SP2KAF9RF86PVX3NEE27DFV1CQX0T4WGR41X3S45C.bitcoin-whales"
```

]

```

> From there, clarinet will be able to reconciliate the `contract-call` present in your local contracts, download and cache a copy of the required contracts, and use them during the execution of your tests suites, and all the different features available in `clarinet`.

Reviewer: I'm not sure what you mean by "clarinet will be able to reconciliate the `contract-call` present in your local contracts" here. Can you word that differently?

Author: Thanks, any take on https://github.com/hirosystems/clarinet/pull/333/commits/c03549d1fa3804faa9fbd19f01e0933d0051ae0f?

> From there, clarinet will be able to reconciliate the `contract-call` present in your local contracts, download and cache a copy of the required contracts, and use them during the execution of your tests suites, and all the different features available in `clarinet`.

Reviewer: ```suggestion
From there, clarinet will be able to reconciliate the `contract-call` present in your local contracts, download and cache a copy of the required contracts, and use them during the execution of your testsuites, and all the different features available in `clarinet`.
```

When deploying your protocol to Devnet / Testnet, for the contracts involving requirements, the setting `remap_requirements` in your deployment plans must be set. 

Before Devnet / Testnet deployments, your contracts will be automatically remapped on the fly to point to the duplicated requirements deployed by an account that you control.

### Deploy contracts to Devnet / Testnet / Mainnet

> You can use Clarinet to publish your contracts to the public Testnet environment for testing and evaluation on a blockchain. 

Reviewer: This sentence mentions publishing to testnet, but the commands below are for publishing to mainnet.

Author: Addressed with https://github.com/hirosystems/clarinet/pull/333/commits/c03549d1fa3804faa9fbd19f01e0933d0051ae0f

The first step is to generate a deployment plan, with the following command: 

```bash
$ clarinet deployment generate --mainnet
```

> After **cautiously** reviewing (and updating if needed) the generated plan, you can use the command:

Reviewer: Will there be documentation about the contents of the deployment manifest somewhere?

Author: This should be straightforward but I can create an issue for this.
